tests/test_ml_tools.py

```diff
@@ -0,0 +1,119 @@
+import awkward as ak
+import dask_awkward as dak
+import numpy as np
+
+import coffea.ml_tools
+
+
+def prepare_jets_array():
+    # Creating jagged Jet-with-constituent array
+    NJETS = 2000
+    NFEAT = 100
+    jets = ak.zip(
+        {
+            "pt": ak.from_numpy(np.random.random(size=NJETS)),
+            "eta": ak.from_numpy(np.random.random(size=NJETS)),
+            "phi": ak.from_numpy(np.random.random(size=NJETS)),
+            "ntrk": ak.from_numpy(np.random.randint(1, 50, size=NJETS)),
+        }
+    )
+    tracks = ak.zip(
+        {
+            "pt": ak.from_regular(np.random.random(size=(NJETS, NFEAT))),
+            "eta": ak.from_regular(np.random.random(size=(NJETS, NFEAT))),
+            "phi": ak.from_regular(np.random.random(size=(NJETS, NFEAT))),
+            "ip2d": ak.from_regular(np.random.random(size=(NJETS, NFEAT))),
+            "ipz": ak.from_regular(np.random.random(size=(NJETS, NFEAT))),
+        }
+    )
+
+    idx = ak.local_index(tracks.pt, axis=-1)
+    tracks = tracks[idx < jets.ntrk]
+    jets["tracks"] = tracks[:]
+
+    ak_jets = jets[:]
+    ak.to_parquet(jets, "ml_tools.parquet")
+    dak_jets = dak.from_parquet("ml_tools.parquet")
+    return ak_jets, dak_jets
+
+
+def triton_testing():
+    # Defining custom wrapper function with awkward padding requirements.
+    class triton_wrapper_test(coffea.ml_tools.triton_wrapper):
+        def awkward_to_numpy(self, jets):
+            def my_pad(arr):
+                return ak.fill_none(ak.pad_none(arr, 100, axis=1, clip=True), 0.0)
+
+            fmap = {
+                "points__0": {
+                    "deta": my_pad(jets.eta - jets.tracks.eta),
+                    "dphi": my_pad(jets.phi - jets.tracks.phi),
+                },
+                "features__1": {
+                    "dr": my_pad(
+                        np.sqrt(
+                            (jets.eta - jets.tracks.eta) ** 2
+                            + (jets.phi - jets.tracks.phi) ** 2
+                        )
+                    ),
+                    "lpt": my_pad(np.log(jets.tracks.pt)),
+                    "lptf": my_pad(
+                        np.log(jets.tracks.pt / ak.sum(jets.tracks.pt, axis=-1))
+                    ),
+                    "ip2d": my_pad(
+                        np.sign(jets.tracks.ip2d) * np.log(jets.tracks.ip2d + 1)
+                    ),
+                    "ipz": my_pad(
+                        np.sign(jets.tracks.ipz) * np.log(jets.tracks.ipz + 1)
+                    ),
+                },
+                "mask__2": {
+                    "mask": my_pad(ak.ones_like(jets.tracks.pt)),
+                },
+            }
+
+            return {
+                k: ak.concatenate(
+                    [x[:, np.newaxis, :] for x in fmap[k].values()], axis=1
+                ).to_numpy()
+                for k in fmap.keys()
+            }
+
+        def dask_touch(self, jets):
+            jets.eta.layout._touch_data(recursive=False)
```

Comment:
Currently, no matter what I do in this dask_touch method, the `dask_awkward.necessary_columns` still shows `jets.pt` as a necessary column. Not sure if this is a dask_awkward issue or wrong handling in the `triton_wrapper.__call_dask__` method.

Reply:
I'll take a look at this in the coming week and see what's up.

Reply:
Ah, I left the `map_partitions(opt_touch_all=True)`.... Simple fix

Reply:
Yes, just confirmed that removing the `opt_touch_all=True` arguments fixes the necessary column issue. However, that means if I missed a field in the `dask_touch`, dask awkward seg faults. We probably want the failure mode to be more descriptive (probably something to be filed upstream)?

Reply:
Yeah - this gets a bit touchy since it's probably not the best to expose users directly to the data-touching interface.

Maybe what can be done is something like:
- a function where the user prepares data for the ML model solely through awkward operations (no to-numpy)
  - therefore no need to worry about type tracers
- from that function the user passes the *separated* stack of arrays in some dict that we stack in automated way
  - {"column_group1": [a, b, c, ...], "columns_group2":, [x, y, z, ...], ...}

Then we can do the touching on those awkward arrays which are guaranteed to be necessary?

Reply:
In preparation for this, I've promoted the `awkward_to_numpy` -> `numpy_evaluate` -> `numpy_to_awkward` calculation routine to a new base class in the `helper.py` file. 

So what we are saying is basically we split the `awkward_to_numpy` method into a `awkward_to_stack` + `stack_to_numpy` method, where the `stack_to_numpy` method is something automatic, and also tallies the arrays to be touched for dask operation? 

Not sure how general this needs to be.

Reply:
My main goal here is to hide from users an interface that's not (yet) publicly exposed and may change often. That kind of thing should be the responsibility of library developers. I think it's something we can automate to the point of hiding it from users. 

Of course, if it gets to clunky trying to hide we, we'll have to expose it. We should, however, try out a few ways of hiding it.

Reply:
With the existing interface, an "easy" middle ground solution would probably to have a function that returns all the required fields: 

```python
dask_touch(jets): 
   return [jets.field1, jets.field2, jets.nested.field1, jets.nested.field2]
```

Then have the function in the typetrace section iterate over the function return and touch these fields. It's one more function book-keeping that is required of the user (and with dask_awkward as it is right now this, will unfortunately hard segfault if the user misses anything). If I cannot come up with an easy to interface and general solution for stacking/ak-to-np, this might be the route we need to take…

Reply:
Changing to the have users overload a `dask_columns` method and return columns of interest, instead of having the users explicitly call `_touch_data`.

This as chosen over the `stack_to_numpy` method, as I found the `awkward.concatenate` has different behavior for normal awkward arrays and awkward arrays used for typetracer.

Reply:
If you haven't already you may want to check `ak.concatenate` again with the most recent release of awkward. It changed a bit.

Reply:
This was tested with awkward==2.1.4 and dask_awkward==main (as of today), and I am still seeing this issue. Right now, having the user implement a `dask_columns` used (can always be udpated in future updates)

Reply:
OK - I'll try my hand at it but it might be a little bit before I have time.

Reply:
@yimuchen For now just get tests running and xgboost in. Then we can get the finishing touches.